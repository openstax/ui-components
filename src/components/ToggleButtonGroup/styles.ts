import styled from "styled-components";
import { ToggleButtonGroup, ToggleButton } from "react-aria-components";
import { colors } from "../../theme";
import { tabListBaseCss } from "../Tabs";

export const StyledToggleButtonGroup = styled(ToggleButtonGroup)`
  ${tabListBaseCss}
`;

export const StyledToggleButton = styled(ToggleButton)`
  all: unset;
  height: 3.6rem;
  padding: 0 0.8rem;
  background-color: ${colors.palette.white};
  border: solid 0.1rem ${colors.palette.pale};
  font-weight: 400;
  font-size: 1.6rem;
  line-height: 2rem;
  white-space: nowrap;
  user-select: none;

  & + & {
    border-left: none;
  }

  &[data-selected] {
    background-color: ${colors.palette.neutralLighter};
  }

  &[data-disabled] {
    color: ${colors.palette.neutralLight};
  }

  &[data-focus-visible] {
    outline: none;
    box-shadow: inset 0 0 0 0.1rem ${colors.palette.black};
  }
`;
