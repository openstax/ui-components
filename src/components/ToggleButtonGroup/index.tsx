import React from "react";
import { StyledToggleButtonGroup, StyledToggleButton } from "./styles";
import { Key } from "react-aria-components";

export interface ToggleButtonGroupProps {
  children: { key: string, value: string }[];
  selectedItems: Set<Key>;
  onSelectionChange?: React.Dispatch<React.SetStateAction<Set<Key>>>;
  selectionMode?: 'single' | 'multiple';
  className?: string;
}

export const ToggleButton = StyledToggleButton;

export const ToggleButtonGroup = (
  {
    children,
    selectedItems,
    onSelectionChange,
    selectionMode = 'single',
    ...props
  }: ToggleButtonGroupProps) => {

  return (
    <StyledToggleButtonGroup
      selectionMode={selectionMode}
      selectedKeys={selectedItems}
      onSelectionChange={onSelectionChange}
      {...props}
    >
      {children.map((child) =>
        <StyledToggleButton
          key={child.key}
          data-key={child.key}
          id={child.key}
          // Allow parents to trigger handlers, works with onPointer events but not with onMouse events
          onPressStart={e => e.continuePropagation()}
        >
          {child.value}
        </StyledToggleButton>
      )}
    </StyledToggleButtonGroup>
  );
}
