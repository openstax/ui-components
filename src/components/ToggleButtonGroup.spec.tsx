import { ToggleButtonGroup } from "./ToggleButtonGroup/index";
import renderer from "react-test-renderer";

describe("ToggleButtonGroup", () => {
  const childrenListWithKeys = [
    { key: "red", value: "Red" },
    { key: "green", value: "Green" },
    { key: "blue", value: "Blue" },
    { key: "yellow", value: "Yellow" },
    { key: "orange", value: "Orange" },
  ];

  it.each`
    selectionMode
    ${"multiple"}
    ${"single"}
    ${undefined}
  `(
    `matches snapshot with selectionMode #selectionMode`,
    ({ selectionMode }) => {
      const tree = renderer
        .create(
          <ToggleButtonGroup
            selectionMode={selectionMode}
            selectedItems={new Set(["red"])}
          >
            {childrenListWithKeys}
          </ToggleButtonGroup>,
        )
        .toJSON();
      expect(tree).toMatchSnapshot();
    },
  );

  it("calls onPressStart and continues propagation", () => {
    const mockEvent = { continuePropagation: jest.fn() };
    const component = renderer.create(
      <ToggleButtonGroup
        selectionMode="single"
        selectedItems={new Set(["red"])}
      >
        {childrenListWithKeys}
      </ToggleButtonGroup>
    );
    const button = component.root.findByProps({ id: "red" });
    button.props.onPressStart(mockEvent);
    expect(mockEvent.continuePropagation).toHaveBeenCalled();
  });
});
