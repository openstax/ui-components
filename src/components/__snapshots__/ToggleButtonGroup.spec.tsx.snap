// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ToggleButtonGroup matches snapshot with selectionMode #selectionMode 1`] = `
<div
  aria-orientation="horizontal"
  className="sc-gsnTZi hmfeGJ"
  data-orientation="horizontal"
  data-rac=""
  onBlurCapture={[Function]}
  onFocusCapture={[Function]}
  onKeyDownCapture={[Function]}
  role="toolbar"
>
  <button
    aria-pressed={true}
    className="sc-dkzDqf gfcEfH"
    data-key="red"
    data-rac=""
    data-selected={true}
    disabled={false}
    onBlur={[Function]}
    onClick={[Function]}
    onDragStart={[Function]}
    onFocus={[Function]}
    onKeyDown={[Function]}
    onMouseDown={[Function]}
    onMouseEnter={[Function]}
    onMouseLeave={[Function]}
    onMouseUp={[Function]}
    onTouchCancel={[Function]}
    onTouchEnd={[Function]}
    onTouchMove={[Function]}
    onTouchStart={[Function]}
    type="button"
  >
    Red
  </button>
  <button
    aria-pressed={false}
    className="sc-dkzDqf gfcEfH"
    data-key="green"
    data-rac=""
    disabled={false}
    onBlur={[Function]}
    onClick={[Function]}
    onDragStart={[Function]}
    onFocus={[Function]}
    onKeyDown={[Function]}
    onMouseDown={[Function]}
    onMouseEnter={[Function]}
    onMouseLeave={[Function]}
    onMouseUp={[Function]}
    onTouchCancel={[Function]}
    onTouchEnd={[Function]}
    onTouchMove={[Function]}
    onTouchStart={[Function]}
    type="button"
  >
    Green
  </button>
  <button
    aria-pressed={false}
    className="sc-dkzDqf gfcEfH"
    data-key="blue"
    data-rac=""
    disabled={false}
    onBlur={[Function]}
    onClick={[Function]}
    onDragStart={[Function]}
    onFocus={[Function]}
    onKeyDown={[Function]}
    onMouseDown={[Function]}
    onMouseEnter={[Function]}
    onMouseLeave={[Function]}
    onMouseUp={[Function]}
    onTouchCancel={[Function]}
    onTouchEnd={[Function]}
    onTouchMove={[Function]}
    onTouchStart={[Function]}
    type="button"
  >
    Blue
  </button>
  <button
    aria-pressed={false}
    className="sc-dkzDqf gfcEfH"
    data-key="yellow"
    data-rac=""
    disabled={false}
    onBlur={[Function]}
    onClick={[Function]}
    onDragStart={[Function]}
    onFocus={[Function]}
    onKeyDown={[Function]}
    onMouseDown={[Function]}
    onMouseEnter={[Function]}
    onMouseLeave={[Function]}
    onMouseUp={[Function]}
    onTouchCancel={[Function]}
    onTouchEnd={[Function]}
    onTouchMove={[Function]}
    onTouchStart={[Function]}
    type="button"
  >
    Yellow
  </button>
  <button
    aria-pressed={false}
    className="sc-dkzDqf gfcEfH"
    data-key="orange"
    data-rac=""
    disabled={false}
    onBlur={[Function]}
    onClick={[Function]}
    onDragStart={[Function]}
    onFocus={[Function]}
    onKeyDown={[Function]}
    onMouseDown={[Function]}
    onMouseEnter={[Function]}
    onMouseLeave={[Function]}
    onMouseUp={[Function]}
    onTouchCancel={[Function]}
    onTouchEnd={[Function]}
    onTouchMove={[Function]}
    onTouchStart={[Function]}
    type="button"
  >
    Orange
  </button>
</div>
`;

exports[`ToggleButtonGroup matches snapshot with selectionMode #selectionMode 2`] = `
<div
  aria-orientation="horizontal"
  className="sc-gsnTZi hmfeGJ"
  data-orientation="horizontal"
  data-rac=""
  onBlurCapture={[Function]}
  onFocusCapture={[Function]}
  onKeyDownCapture={[Function]}
  role="radiogroup"
>
  <button
    aria-checked={true}
    className="sc-dkzDqf gfcEfH"
    data-key="red"
    data-rac=""
    data-selected={true}
    disabled={false}
    onBlur={[Function]}
    onClick={[Function]}
    onDragStart={[Function]}
    onFocus={[Function]}
    onKeyDown={[Function]}
    onMouseDown={[Function]}
    onMouseEnter={[Function]}
    onMouseLeave={[Function]}
    onMouseUp={[Function]}
    onTouchCancel={[Function]}
    onTouchEnd={[Function]}
    onTouchMove={[Function]}
    onTouchStart={[Function]}
    role="radio"
    type="button"
  >
    Red
  </button>
  <button
    aria-checked={false}
    className="sc-dkzDqf gfcEfH"
    data-key="green"
    data-rac=""
    disabled={false}
    onBlur={[Function]}
    onClick={[Function]}
    onDragStart={[Function]}
    onFocus={[Function]}
    onKeyDown={[Function]}
    onMouseDown={[Function]}
    onMouseEnter={[Function]}
    onMouseLeave={[Function]}
    onMouseUp={[Function]}
    onTouchCancel={[Function]}
    onTouchEnd={[Function]}
    onTouchMove={[Function]}
    onTouchStart={[Function]}
    role="radio"
    type="button"
  >
    Green
  </button>
  <button
    aria-checked={false}
    className="sc-dkzDqf gfcEfH"
    data-key="blue"
    data-rac=""
    disabled={false}
    onBlur={[Function]}
    onClick={[Function]}
    onDragStart={[Function]}
    onFocus={[Function]}
    onKeyDown={[Function]}
    onMouseDown={[Function]}
    onMouseEnter={[Function]}
    onMouseLeave={[Function]}
    onMouseUp={[Function]}
    onTouchCancel={[Function]}
    onTouchEnd={[Function]}
    onTouchMove={[Function]}
    onTouchStart={[Function]}
    role="radio"
    type="button"
  >
    Blue
  </button>
  <button
    aria-checked={false}
    className="sc-dkzDqf gfcEfH"
    data-key="yellow"
    data-rac=""
    disabled={false}
    onBlur={[Function]}
    onClick={[Function]}
    onDragStart={[Function]}
    onFocus={[Function]}
    onKeyDown={[Function]}
    onMouseDown={[Function]}
    onMouseEnter={[Function]}
    onMouseLeave={[Function]}
    onMouseUp={[Function]}
    onTouchCancel={[Function]}
    onTouchEnd={[Function]}
    onTouchMove={[Function]}
    onTouchStart={[Function]}
    role="radio"
    type="button"
  >
    Yellow
  </button>
  <button
    aria-checked={false}
    className="sc-dkzDqf gfcEfH"
    data-key="orange"
    data-rac=""
    disabled={false}
    onBlur={[Function]}
    onClick={[Function]}
    onDragStart={[Function]}
    onFocus={[Function]}
    onKeyDown={[Function]}
    onMouseDown={[Function]}
    onMouseEnter={[Function]}
    onMouseLeave={[Function]}
    onMouseUp={[Function]}
    onTouchCancel={[Function]}
    onTouchEnd={[Function]}
    onTouchMove={[Function]}
    onTouchStart={[Function]}
    role="radio"
    type="button"
  >
    Orange
  </button>
</div>
`;

exports[`ToggleButtonGroup matches snapshot with selectionMode #selectionMode 3`] = `
<div
  aria-orientation="horizontal"
  className="sc-gsnTZi hmfeGJ"
  data-orientation="horizontal"
  data-rac=""
  onBlurCapture={[Function]}
  onFocusCapture={[Function]}
  onKeyDownCapture={[Function]}
  role="radiogroup"
>
  <button
    aria-checked={true}
    className="sc-dkzDqf gfcEfH"
    data-key="red"
    data-rac=""
    data-selected={true}
    disabled={false}
    onBlur={[Function]}
    onClick={[Function]}
    onDragStart={[Function]}
    onFocus={[Function]}
    onKeyDown={[Function]}
    onMouseDown={[Function]}
    onMouseEnter={[Function]}
    onMouseLeave={[Function]}
    onMouseUp={[Function]}
    onTouchCancel={[Function]}
    onTouchEnd={[Function]}
    onTouchMove={[Function]}
    onTouchStart={[Function]}
    role="radio"
    type="button"
  >
    Red
  </button>
  <button
    aria-checked={false}
    className="sc-dkzDqf gfcEfH"
    data-key="green"
    data-rac=""
    disabled={false}
    onBlur={[Function]}
    onClick={[Function]}
    onDragStart={[Function]}
    onFocus={[Function]}
    onKeyDown={[Function]}
    onMouseDown={[Function]}
    onMouseEnter={[Function]}
    onMouseLeave={[Function]}
    onMouseUp={[Function]}
    onTouchCancel={[Function]}
    onTouchEnd={[Function]}
    onTouchMove={[Function]}
    onTouchStart={[Function]}
    role="radio"
    type="button"
  >
    Green
  </button>
  <button
    aria-checked={false}
    className="sc-dkzDqf gfcEfH"
    data-key="blue"
    data-rac=""
    disabled={false}
    onBlur={[Function]}
    onClick={[Function]}
    onDragStart={[Function]}
    onFocus={[Function]}
    onKeyDown={[Function]}
    onMouseDown={[Function]}
    onMouseEnter={[Function]}
    onMouseLeave={[Function]}
    onMouseUp={[Function]}
    onTouchCancel={[Function]}
    onTouchEnd={[Function]}
    onTouchMove={[Function]}
    onTouchStart={[Function]}
    role="radio"
    type="button"
  >
    Blue
  </button>
  <button
    aria-checked={false}
    className="sc-dkzDqf gfcEfH"
    data-key="yellow"
    data-rac=""
    disabled={false}
    onBlur={[Function]}
    onClick={[Function]}
    onDragStart={[Function]}
    onFocus={[Function]}
    onKeyDown={[Function]}
    onMouseDown={[Function]}
    onMouseEnter={[Function]}
    onMouseLeave={[Function]}
    onMouseUp={[Function]}
    onTouchCancel={[Function]}
    onTouchEnd={[Function]}
    onTouchMove={[Function]}
    onTouchStart={[Function]}
    role="radio"
    type="button"
  >
    Yellow
  </button>
  <button
    aria-checked={false}
    className="sc-dkzDqf gfcEfH"
    data-key="orange"
    data-rac=""
    disabled={false}
    onBlur={[Function]}
    onClick={[Function]}
    onDragStart={[Function]}
    onFocus={[Function]}
    onKeyDown={[Function]}
    onMouseDown={[Function]}
    onMouseEnter={[Function]}
    onMouseLeave={[Function]}
    onMouseUp={[Function]}
    onTouchCancel={[Function]}
    onTouchEnd={[Function]}
    onTouchMove={[Function]}
    onTouchStart={[Function]}
    role="radio"
    type="button"
  >
    Orange
  </button>
</div>
`;
